plio/views.py

```diff
@@ -451,79 +451,94 @@ def metrics(self, request, uuid):
+                for group in user_grouping.groups:
+                    # get the responses for a given user
+                    group_df = user_grouping.get_group(group)
 
-            # only use the responses from viewers who have answered at least
-            # one question to compute the accuracy
-            answered_at_least_one_index = num_answered_list > 0
-            num_answered_list = num_answered_list[answered_at_least_one_index]
-            num_correct_list = num_correct_list[answered_at_least_one_index]
+                    num_answered = sum(
+                        group_df["answer"].apply(lambda value: value is not None)
+                    )
 
-            if not len(num_correct_list):
-                accuracy = None
-            else:
-                accuracy = np.round(
-                    (num_correct_list / num_answered_list).mean() * 100, 2
+                    num_answered_list.append(num_answered)
+
+                    if not num_answered:
+                        num_correct_list.append(None)
+                    else:
+                        num_correct_list.append(
+                            sum(group_df.apply(is_answer_correct, axis=1))
+                        )
+
+                # converting to numpy arrays enabled us to use vectorization
+                # to speed up the computation many folds
+                num_answered_list = np.array(num_answered_list)
+                num_correct_list = np.array(num_correct_list)
+                average_num_answered = round(num_answered_list.mean())
+                percent_completed = np.round(
+                    100
+                    * (sum(num_answered_list == num_questions) / num_unique_viewers),
+                    2,
                 )
 
-        return Response(
-            {
-                "unique_viewers": num_unique_viewers,
-                "average_watch_time": average_watch_time,
-                "percent_one_minute_retention": percent_one_minute_retention,
-                "accuracy": accuracy,
-                "average_num_answered": average_num_answered,
-                "percent_completed": percent_completed,
-            }
-        )
+                # only use the responses from viewers who have answered at least
+                # one question to compute the accuracy
+                answered_at_least_one_index = num_answered_list > 0
+                num_answered_list = num_answered_list[answered_at_least_one_index]
+                num_correct_list = num_correct_list[answered_at_least_one_index]
+
+                if not len(num_correct_list):
+                    accuracy = None
+                else:
+                    accuracy = np.round(
+                        (num_correct_list / num_answered_list).mean() * 100, 2
+                    )
+
+                if num_questions != len(dff):
+                    quesSurvey = (True,)
+                else:
+                    quesSurvey = (False,)
```

Comment:
you can use a simpler and easier to read check: (question_df['survey'] == True).sum() > 0

Reply:
The updated question_df does not have any `survey=True` so this will not work.